Feature/udp server client

Ian-Carroll opened the pull request:
Just implemented the UDPServer and UDPClient classes with some basic socket code!

Ian-Carroll committed:
Roughed in UDPServer & UDPClient classes.

Ian-Carroll committed:
Implemented the open(), close(), write(), and read(), methods for UDPServer & UDPClient classes.

Ian-Carroll committed:
Moved the SerialSettings class and SerialComms to serial.py.

Ian-Carroll committed:
Moved the UDPServer class and UDPClient class into udp.py.

Ian-Carroll committed:
Removed UDP classes and Serial classes from comms.py.

Burke-Daniel committed:
Create base device class and other primitives

Includes functions for constructing devices,
constructing comms objects, running device processes,
and a derived StreamingDevice class which is to be used
for devices which stream static data at a fixed interval.

Also create sample device_config.json to get idea for what
config file structure might look like.

Burke-Daniel reviewed (changes requested):
So I was just trying to throw together a simple example to test this before giving it the stamp of approval, but I have found quite a few things that are not working properly so instead of me pointing them all out, I'll get you to go ahead and try to make a simple test with client and server for one to send to the other, and the print the data to console. You should be able to find all the bugs/syntax errors that way!

Ian-Carroll committed:
Implemented a UDPClient for learning under examples.

Ian-Carroll committed:
Implemented a UDPServer for learning under examples.

Ian-Carroll committed:
Changed syntax errors in read and write methods for UDPServer and Client, added functional testing as well.

Ian-Carroll committed:
Implemented testing and appropriate changes on UDPClient & Server classes. Note: Client read function is not working correctly.

Ian-Carroll committed:
Added comms to import path.

Ian-Carroll committed:
Worked examples of UDPClient and UDPServer classes.

Burke-Daniel committed:
Fix bugs found during testing

Dataclass is a decorator, create a default serial object and then
fill in port and baud rate. Also make bug fixes in the device class.

Burke-Daniel committed:
Device class refinements

Remove comms type from base Device class. Base device class now
just contains type and id, and abstract methods.

Add StreamingDeviceSettings to follow design paradigm used in
comms classes.

Update example to use these changes.

Ian-Carroll committed:
Referred functional testing to worked examples of udp.py.

Burke-Daniel commented on examples/UDPClient.py:
```diff
@@ -0,0 +1,10 @@
+data, addr = client_sock.recvfrom(4096)
+print("Server Says")
+print(str(data))
+client_sock.close()
```
Newline

Burke-Daniel commented on testing_platform/device/udp.py:
```diff
@@ -0,0 +1,50 @@
+    def read(self, num_bytes):
+        self.sock.recvfrom(num_bytes)
+        data,addr = self.sock.recvfrom(num_bytes)
+        print(str(data))
+        print(str(addr))
```
Remove these print statements and I'd say just return the data from this function

Burke-Daniel commented on testing_platform/device/udp.py:
```diff
@@ -0,0 +1,50 @@
+    def read (self, num_bytes):
+        self.sock.recvfrom(num_bytes)
+        data,addr = self.sock.recvfrom(num_bytes)
+        print(str(data))
+        print(str(addr))
```
Same thing as above

Burke-Daniel commented on examples/UDPServer.py:
```diff
@@ -0,0 +1,11 @@
+    print(str(data))
+    print(str(addr))
+    msg = b"Hello, I am a UDP Server!"
+    sock.sendto(msg, addr)
```
This never returns, and thus you have to kill your terminal to get it to exit. Maybe look at this for a solution https://www.kite.com/python/answers/how-to-handle-a-sigterm-interrupt-in-python

Burke-Daniel reviewed (changes requested):
All looks good, see my few comments, I would rebase onto my general device class branch so you can integrate UDP in to that class as well.

Burke-Daniel commented on testing_platform/device/udp.py:
```diff
@@ -0,0 +1,50 @@
+        self.sock.close()
+
+    def write(self, data):
+        self.sock.sendto(data, ("127.0.0.1", 12345))
```
Make the address and port configurable

Burke-Daniel commented on testing_platform/device/udp.py:
```diff
@@ -0,0 +1,50 @@
+        self.sock.sendto(data, ("127.0.0.1", 12345))
+
+    def read(self, num_bytes):
+        self.sock.recvfrom(num_bytes)
```
I think you can remove this line

Ian-Carroll committed:
Added configurable port & address

Ian-Carroll committed:
Newline

Ian-Carroll committed:
Added triple quotes for device_config_s.

Ian-Carroll committed:
Implemented an exit condition for UDPServer check loop.

Burke-Daniel commented on README.md:
````diff
@@ -7,7 +7,7 @@ the standard python venv module.
 **Note**: Python 3.8+ is required.
 ```
-cd {REPO_ROOT}
+cd {REPO_LOCATION}/testing-platform
````
Should be {REPO_LOCATION}/testing_platform I believe

Burke-Daniel commented on examples/UDPClient.py:
```diff
@@ -0,0 +1,10 @@
+data, addr = client_sock.recvfrom(4096)
+print("Server Says")
+print(str(data))
+client_sock.close()
```
Just to be proper, I would through this code in a function called main(), and then call that function under `if __name__ == '__main__'`

Burke-Daniel commented on examples/UDPServer.py:
```diff
@@ -0,0 +1,17 @@
+        sock.sendto(msg, addr)
+except KeyboardInterrupt:
+    print("\nCtrl+C terminated the loop.")
+    pass
```
Same comment as in UDPServer.py

Burke-Daniel commented on examples/example.py:
```diff
@@ -1 +1 @@
-# Intentionally left blank
+# Intentionally left blank
```
You're good to delete this file with `git rm`

Burke-Daniel commented on examples/UDPServer.py:
```diff
@@ -0,0 +1,17 @@
+import socket
+import keyboard
```
Can remove this

Burke-Daniel reviewed (approved):
Couples comments for you to address then GTM!

Ian-Carroll committed:
Deleted example.py

Ian-Carroll committed:
Fixed repo location.

Ian-Carroll committed:
Polished UDP examples.

Ian-Carroll committed:
Merge
